### ISSUE TITLE
[suggestion] Color-neutral expedition normal success/failure indicator

### ISSUE THREAD

**Issue by transientinfinity** [KC3Kai/KC3Kai#2263]
Not a traditional bug, but a suggestion:

tl;dr: Suggest adding some sort of non-color-based indicator for probable/certain failure of normal expeditions, for a more colorblind friendly UI. (Or, heck, even a colorblind mode that exposes the styles for color-coded text/backgrounds so users can easily tweak them according to their personal needs. E.g., fleet tagging in the ship-list can be a bit hard to differentiate at times.)

Currently the success or failure of normal expeditions is indicated by the color coding of various pre-conditions—fleet comp, flagship type and level, total fleet level, et cetera. Whereas great success chances are indicated by a % indicator—86% chance, 96% chance, et cetera. For great success this works very well, as you can see immediately from the panel if there's still something you need to do. If you have insufficient drums then great success chance will be zero, insufficient sparkle leads to sub-100% great success chance, et cetera. For normal, however, the only indicator is when pre-condition listings are red (at least, I think they're red). If you absentmindedly send out a fleet w/ a level 2 flagship, for example (\*cough\*never done this\*cough\*), expedition 6 will fail.

It's a minor thing, but I'd suggest possibly adding a text-only indicator that shows up in the case that a normal expedition will fail for the currently selected fleet, similar to how "(X) needs supply" replaces "(check mark) supply good". It could be as simple as "unmet conditions", since that will signal to the user that they need to take a closer look at fleet comp. Or even just a check mark next to met conditions, making it a more literal checklist with UI cues reminding the user to use it as such. Or it could copy the style of the great-success variant of the panel (reached by clicking the icon) and just say "0% normal success"/"100% normal success" based on the various conditions. There are almost certainly even more imaginative/elegant/simple ways it could be approached.

Yes, there's the automatic "well maybe you shouldn't be so stupid" response to be expected from forum trolls, but the fact of the matter is that the UI *already* contains these indicators in the form of pre-condition color-coding. Sure, the user will eventually learn to check their fleet comp each and every time they make a change before semi-automatically sending their fleets back out, but I think it would be a user-friendly addition for the estimated 5-10% of men and 0.5-1% of women affected by colorblindness.

That's, like, an expected 5% improvement in overall user satisfaction! And all it takes is an extra two words! (#ThatsNotHowCodingActuallyWorks #ButYeah)

Great job, btw. This extension adds so much QoL to the KC experience.

**Comment by nonsensed** [comment]
We were just having this discussion in the kc3kai discord the other day. As someone who is red-green colorblind myself, I know what you mean. I've just been dealing with it and wasn't going to do anything if no one else reported the issue.

Personally, I was of the opinion that adding a green that contrasts better would help a lot with this.

![img](https://i.imgur.com/OCHJmiF.png)
This for me is the biggest problem. See the 37 and 38? It doesn't look different at all from the 35, 36, etc.

![img](https://i.imgur.com/3JVgv1D.png)
But same here-- the red and green are difficult to tell apart.

If I have time, I'm going to try to play with the CSS later and see if I can come up with something that looks good for cb and non-cb people, but if someone else wants to, go right ahead.

Another thing that might help is adding a `FAILURE` icon to go along with the `GREAT` and `SUCCESS` icons.

**Comment by transientinfinity** [comment]
Assuming the cell backgrounds can be individually styled, a low-brightness/saturation and high-brightness/saturation highlight for the expedition title—a sort of budget "sparkle" effect—could also be a more obvious and color-neutral option.

If you'll forgive the MS Paint skills, it might look something like this (using a modified version of your original image to demonstrate the difference):

![68747470733a2f2f692e696d6775722e636f6d2f4f43484a6d69462e706e67_2](https://user-images.githubusercontent.com/19354299/31255548-6765d5b0-a9fb-11e7-948b-56c90c27a12f.png)

**Comment by nonsensed** [comment]
Opinions? Compromises?
![img](https://i.imgur.com/lvVzlUw.png)

![img](https://i.imgur.com/p7i4NZJ.png)
Too red?

**Comment by transientinfinity** [comment]
The red seems to pop out a bit more, which is good. Though it's hard to really tell how effective it might be since the image is a bit more zoomed in and I'm specifically looking for differences—larger areas of R/G w/ higher light levels are easier for me to differentiate, if I'm able to differentiate them to begin with.

The backgrounds on the dropdown make it a lot easier to tell offhand if your fleet is ready when selecting expeditions, though your clip doesn't make it clear if the non-dropdown cell would also be faded out like that. That said, the fact that the dropdown has contrasting backgrounds might make a shaded *main* title a less obvious indicator, anyway.

Underlining conditions that still need to be satisfied might help draw attention to them without being too intrusive, though I rarely actually see underlining used anywhere online/IRL—maybe italics? A span based background highlight could also work, though that might be too "in your face". Then again, maybe "in your face" is what you want when reminding the user to check their fleet conditions, as long as it isn't "taiha heartbeat" levels of annoying. I don't really have any professional experience in graphic design or UX, so I'm not sure what sort of balance is desirable.

**Comment by sinsinpub** [comment]
the best way should design 3 more themes (color schemes) for color blindness (`Protanopia` `Deuteranopia` `Tritanopia` even `Monochromatic`). but seems we have no such dev can achieve this... 😓 

solutions i guess could be feasibility:
 * find a tool to auto convert our css files for general types, allow switching via settings. i googled a thing: https://www.npmjs.com/package/postcss-colorblind
 * find those elements hard to be recognized by you, we try to offer the custom css codes and finally get a useful css list which are really fit of your type.

**Comment by transientinfinity** [comment]
Dealing w/ the various kinds of colorblindness does seem like it could get pretty messy, which is part of why I was suggesting a color-neutral solution using other indicators—brightness, highlighting, font-effects, et cetera non color-based solutions. Custom CSS would certainly provide a solution, but on your end I would expect that needing to add additional elements to the settings interface in order to expose them would be more complicated than making a modification to any existing CSS styles you may be using.

//fake french accent:
"Three hours later..."

After digging around for a bit, here's an example that someone could load into the CSS customization for the panel.

    /*expedition id list*/
    .module.activity .activity_expeditionPlanner .expedition_entry.cond_failed {
    text-decoration: line-through;
    background-color: #ccc;
    }

    /*expedition-specific requirements*/
    .module.activity .activity_expeditionPlanner .expPlanner_text_failed {
    text-decoration: line-through;
    }

... granted, now that I've figured that out, I guess I don't really "need" to request such a feature. Nonetheless, it does seem like it would be a useful thing to address for other potentially colorblind users, even if it's just by adding a note to one of the "learn more" or "help topics" sections that instructs the user what CSS classes get applied to failed expedition requirements, without having to dig through the (in some cases apparently obfuscated?) source. It'd probably go faster for someone less easily confused/distracted, but it still seems like a bit of work if you have no idea what you're looking for going in.

**Comment by nonsensed** [comment]
Well, I've taken some time to think about it. I didn't suggest underline or strikethrough because I tried it in my panel and personally I don't like how it looks. Even if we made a panel with that setup, I'm colorblind and I wouldn't even use it.

@sinsinpub It's not necessary to have so many themes, honestly. You can support all of those colorblindness types simply by using a single alternate style that uses high-contrast color schemes. I was trying to support still using red in my example above to appease people who aren't colorblind, but turning the red to black pretty much solves the problem completely for all colorblindness types. The only problem is if people modify their panel opacity to use a very dark picture where the black becomes invisible. But some people already use background images that make their panel text almost invisible... so... xD

I think @transientinfinity solved his own problem to his liking? If no one likes my suggestion, I guess I'll customize my own css (with the red turned into black) and we can be done with this issue?

**Comment by transientinfinity** [comment]
I wouldn't really say "to my liking"—my solution is ugly as sin—but for the moment at least I've addressed it. I still think there should be some sort of colorblind-friendly setup for the general userbase, as the difference can be like night and day, and it's something people just live with so often that they don't even realize how much less efficient things are when your ability to make decisions based on color-coding is impaired.

But, that said, I've been waiting well over a year for Factorio to address their map colorings in the base game (copper or biter hive? only one way to find out!), so I do recognize that it's not really ever considered an issue by developers. In the grand scheme of things, it's a relatively minor QoL improvement that impacts less than 5-10% of users (assuming statistical independence), and w/ HTML5 apparently inbound soon(?)(TM), I'm sure there are other things that need to be done.

**Comment by nonsensed** [comment]
Ah. I'll throw this here, then. Play with it if you want.

```
.module.activity .activity_expeditionPlanner .expPlanner_text.gsrate_content[data-gsState="impossible"] {
	color:black;
}
.module.activity .activity_expeditionPlanner .expPlanner_text.gsrate_content[data-gsState="no-overdrum"] {
	color:orange;
}
.module.activity .activity_expeditionPlanner .expPlanner_text.gsrate_content[data-gsState="likely"] {
	color:#00CC00;
}
.module.activity .activity_expeditionPlanner .expPlanner_text.gsrate_content[data-gsState="guaranteed"] {
	color:gold;
}
.module.activity .activity_expeditionPlanner .expPlanner_text_passed {
	font-size: 14px;
	color:#00CC00;
}
.module.activity .activity_expeditionPlanner .expPlanner_text_failed {
	font-size: 14px;
	color:black;
}
.module.activity .activity_expeditionPlanner .expedition_entry.cond_passed {
	color:00BB00;
}
.module.activity .activity_expeditionPlanner .expedition_entry.cond_failed {
	color:#AA5050;
	opacity: 0.6;
}
```

**Comment by ethrundr** [comment]
@nonsensed could you please make the green one to have darker grey?

While your issue represents people that are colorblind, my own eyes are bit difficult to see contrasts.
Unless it has some definite difference, I can't tell apart which is lighter and which is darker.

Well, it can be compromised tho by lowering the contrast and whitebalance on my screen, but I do think this should stand out without user need to resetup their monitors, where is impossible in certain models.

**Comment by nonsensed** [comment]
In the dropdown or the expedition comp one?

Monitor settings are another issue. There is no such thing as a standard for everyone's monitor calibration. This is very troublesome...

Btw, this was the css I posted above's output:
![img](https://i.imgur.com/x02C8Pf.png)

or a different green (I assume it's green!):
![img](https://i.imgur.com/nLqNnxH.png)

or bring back the red and go say Merry Christmas:
![img](https://i.imgur.com/xz935Mv.png)

...but I don't think anyone wants to use my settings xD

Also I just realized the top Expedition thing is probably supposed to be a color. Oh boy.

**Comment by sinsinpub** [comment]
i guess @ethrundr misread... @nonsensed 's CSS codes are supposed to be applied to their 'Custom CSS' settings, not committed to our repo 😅

**Comment by transientinfinity** [comment]
Ideally some change would eventually be made dev-side, since I doubt the majority of KC3Kai users visit the github that often. Even if they don't realize they'd benefit from a change, I expect they'd appreciate it if they suddenly realized that the success+failure conditions were color/contrast-coded.

That said, IMO the change would have to be something that both addresses potential color-vision/contrast-sensitivity impairments and doesn't tick off other users. E.g. the temp-underlining/strikethrough solution I was using is godsawful ugly, as indicated by @nonsensed's reaction and my own statements. Ultimately I just ended up using the slow-flashing text solution. It's slow enough not to be too distracting, but fast enough to be obvious, with the added benefit that it doesn't obscure the requirements like the strike-through did.

    /*expedition-specific requirements*/
    .module.activity .activity_expeditionPlanner .expPlanner_text_failed {
        animation: blinker 2s linear infinite;
    }

    @keyframes blinker {  
        25% { opacity: 0; }
        50% { opacity: 1; }
    }

Of course, it could also be a toggle like the taiha alerts are, since multiple classes can be applied to a single element. Add a few lines in the relevant CSS files to enable the animation, and an added if statement in the logic that applies and removes the pass/fail styles—plus a note in the patch notes mentioning it, since most of the settings are "touch once and forget". Not now, or this, necessarily, but something, eventually, would potentially be a nice QoL change for those ~1-5% of users. Being optional, consensus on the "but what should it look like" would be slightly less mandatory than a major change in presentation of the UI. Presumably some amount of feedback might come in once the option existed to be noticed, as well, if further tweaking seemed necessary.

I dunno. Just some thoughts.
